packages/lodestar/src/chain/eventHandlers.ts

```diff
@@ -1,48 +1,86 @@
 import {readOnlyMap, toHexString} from "@chainsafe/ssz";
 import {Attestation, Checkpoint, SignedBeaconBlock, Slot} from "@chainsafe/lodestar-types";
-import {toJson} from "@chainsafe/lodestar-utils";
+import {ILogger, toJson} from "@chainsafe/lodestar-utils";
 import {IBlockSummary} from "@chainsafe/lodestar-fork-choice";
 
 import {ITreeStateContext} from "../db/api/beacon/stateContextCache";
 import {AttestationError, AttestationErrorCode, BlockError, BlockErrorCode} from "./errors";
 import {IBlockJob} from "./interface";
-import {IChainEvents} from "./emitter";
+import {ChainEventEmitter, IChainEvents} from "./emitter";
 import {BeaconChain} from "./chain";
 
 interface IEventMap<Events, Key extends keyof Events = keyof Events, Value extends Events[Key] = Events[Key]>
   extends Map<Key, Value> {
   set<Key extends keyof Events>(key: Key, value: Events[Key]): this;
 }
 
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type ListenerType<T> = [T] extends [(...args: infer U) => any] ? U : [T] extends [void] ? [] : [T];
+
+/**
+ * Returns a function that runs an async handler function with input args,
+ * If the handler function successfully completes,
+ * then emits the event on the event emitter using the same args
+ */
+function wrapHandler<
+  Key extends keyof IChainEvents = keyof IChainEvents,
+  Value extends IChainEvents[Key] = IChainEvents[Key]
+>(event: Key, emitter: ChainEventEmitter, logger: ILogger, handler: (...args: Parameters<Value>) => Promise<void>) {
+  return async (...args: Parameters<Value>): Promise<void> => {
+    try {
+      await handler(...args);
+      emitter.emit(event, ...((args as unknown) as ListenerType<Value>));
```

Comment:
It's odd that you have to cast it as unknown, Parameters<Value> should be sufficient

Reply:
For some reason (unknown to me), `Parameters<Value>` wasn't working.
Thats why `ListenerType` is copied here (direct copy-paste from the events type declarations).